File: medium/0912_sort_an_array/solution.cpp
#include <queue>
#include <vector>

class Solution {
public:
  std::vector<int> sortArray(std::vector<int> &nums) {
    std::deque<std::vector<int>> dq;
    for (const auto &num : nums)
      dq.push_back({num});

    while (dq.size() > 1) {
      auto v1 = dq.front();
      dq.pop_front();
      auto v2 = dq.front();
      dq.pop_front();

      size_t n1 = v1.size();
      size_t n2 = v2.size();

      size_t i1{};
      size_t i2{};

      std::vector<int> v3(n1 + n2);

      for (size_t i{}; i < n1 + n2; ++i) {
        if (i1 == n1 || (i2 != n2 && v1[i1] > v2[i2]))
          v3[i] = v2[i2++];
        else
          v3[i] = v1[i1++];
      }

      dq.push_back(v3);
    }

    return dq.front();
  }
};

File: medium/0912_sort_an_array/solution.py
class Solution:
    def sortArray(self, nums: list[int]) -> list[int]:
        n = len(nums)

        def siftDown(i: int) -> None:
            while True:
                left = 2 * i + 1
                right = 2 * i + 2

                largest = i

                if left < n and nums[left] > nums[largest]:
                    largest = left
                if right < n and nums[right] > nums[largest]:
                    largest = right

                if largest == i:
                    break

                nums[i], nums[largest] = nums[largest], nums[i]
                i = largest

        def heapify():
            for i in range(n // 2 - 1, -1, -1):
                siftDown(i)

        heapify()

        for n in range(n - 1, -1, -1):
            nums[0], nums[n] = nums[n], nums[0]
            siftDown(0)

        return nums
